File: nfc_proof_of_concept/lib/app_layer/state_machine.dart
import 'dart:math';
import 'dart:typed_data';

import 'file_access/fields/response_fields.dart';
import 'file_access/serializers/tlv_block_serializer.dart';
import 'file_access/fields/command_fields.dart';
import 'file_access/serializers/apdu_command_serializer.dart';
import 'file_access/serializers/apdu_response_serializer.dart';
import 'file_access/serializers/capability_container_serializer.dart';
import 'ndef_format/ndef_message_serializer.dart';


enum _HceState {
  idle,
  appSelected,
  ccSelected,
  ndefSelected,
}


class _NdefFile {
  late Uint8List _bytes;
  final int maxSize;

  Uint8List get buffer => _bytes;
  int get currentSize => _bytes.length;

  _NdefFile({required NdefMessageSerializer message, required this.maxSize}) {
    update(message);
  }

  void update(NdefMessageSerializer message) {
    final messageBytes = message.buffer;
    final nlen = messageBytes.length;

    if (nlen + 2 > maxSize) {
      throw ArgumentError('NDEF message size ($nlen bytes) exceeds the max file size ($maxSize bytes) defined in the CC.');
    }

    final nlenBytes = Uint8List(2);
    nlenBytes[0] = (nlen >> 8) & 0xFF;
    nlenBytes[1] = nlen & 0xFF;
    
    _bytes = Uint8List.fromList(nlenBytes + messageBytes);
  }
}


class HceStateMachine {
  _HceState _currentState = _HceState.idle;

  final CapabilityContainer capabilityContainer;
  final _NdefFile ndefFile;

  // El AID estándar para la aplicación NDEF.
  static final ndefAid = Uint8List.fromList([0xA0, 0x00, 0xDA, 0xDA, 0xDA, 0xDA, 0xDA]);

  HceStateMachine({
    required NdefMessageSerializer initialMessage,
    bool isWritable = false,
    int maxNdefFileSize = 2048, // 2KB
  }) : capabilityContainer = CapabilityContainer(
          fileDescriptors: [
            FileControlTlv.ndef(
              maxNdefFileSize: maxNdefFileSize,
              isNdefWritable: isWritable,
            ),
          ],
        ),
        ndefFile = _NdefFile(message: initialMessage, maxSize: maxNdefFileSize);

  /// Resets the state machine to its initial state.
  /// Call this when the NFC field is deactivated.
  void onDeactivated() {
    _currentState = _HceState.idle;
  }

  /// The main entry point for processing an incoming APDU command.
  /// Takes a raw command and returns a raw response.
  Uint8List processCommand(Uint8List rawCommand) {
    try {
      final command = ApduCommand.fromBytes(rawCommand);
      final response = _handleCommand(command);
      return response.buffer;
    } on ArgumentError catch (e) {
      // Error de parseo (trama malformada)
      print("APDU Parsing Error: $e");
      return ApduResponse.error(ApduStatusWord.wrongLength).buffer;
    } on UnimplementedError catch (e) {
      // INS no soportado
      print("Unsupported Instruction: $e");
      return ApduResponse.error(ApduStatusWord.insNotSupported).buffer;
    } catch (e) {
      // Error inesperado
      print("Unexpected FSM Error: $e");
      return ApduResponse.error(ApduStatusWord.conditionsNotSatisfied).buffer;
    }
  }

  /// The core FSM logic. Routes commands based on the current state.
  ApduResponse _handleCommand(ApduCommand command) {
    switch (_currentState) {
      case _HceState.idle:
        return _handleIdleState(command);
      case _HceState.appSelected:
        return _handleAppSelectedState(command);
      case _HceState.ccSelected:
        return _handleCcSelectedState(command);
      case _HceState.ndefSelected:
        return _handleNdefSelectedState(command);
    }
  }


  ApduResponse _handleIdleState(ApduCommand command) {
    if (command is SelectCommand) {
      if (command.params == ApduParams.byName && _isNdefAid(command.data.buffer)) {
        _currentState = _HceState.appSelected;
        return ApduResponse.success();
      }
    }
    return ApduResponse.error(ApduStatusWord.conditionsNotSatisfied);
  }

  ApduResponse _handleAppSelectedState(ApduCommand command) {
    if (command is SelectCommand) {
      // En APP_SELECTED, solo aceptamos SELECT by File ID
      if (command.params == ApduParams.byFileId) {
        final fileId = (command.data!.buffer[0] << 8) | command.data!.buffer[1];
        if (fileId == 0xE103) { // CC File
          _currentState = _HceState.ccSelected;
          return ApduResponse.success();
        } else if (fileId == 0xE104) { // NDEF File
          _currentState = _HceState.ndefSelected;
          return ApduResponse.success();
        } else {
          return ApduResponse.error(ApduStatusWord.fileNotFound);
        }
      }
    }
    return ApduResponse.error(ApduStatusWord.conditionsNotSatisfied);
  }

  ApduResponse _handleCcSelectedState(ApduCommand command) {
    if (command is ReadBinaryCommand) {
      return _processRead(command, capabilityContainer.buffer);
    }
    return ApduResponse.error(ApduStatusWord.insNotSupported);
  }

  ApduResponse _handleNdefSelectedState(ApduCommand command) {
    if (command is ReadBinaryCommand) {
      return _processRead(command, ndefFile.buffer);
    }

    return ApduResponse.error(ApduStatusWord.insNotSupported);
  }


  ApduResponse _processRead(ReadBinaryCommand command, Uint8List file) {
    final offset = command.offset;
    if (offset >= file.length) {
      return ApduResponse.error(ApduStatusWord.wrongOffset);
    }
    
    final lengthToRead = command.lengthToRead == 0 ? 256 : command.lengthToRead;
    final bytesRemaining = file.length - offset;
    final bytesToSend = min(lengthToRead, bytesRemaining);

    final chunk = file.sublist(offset, offset + bytesToSend);
    return ApduResponse.success(data: chunk);
  }

  bool _isNdefAid(Uint8List? aid) {
    if (aid == null || aid.length != ndefAid.length) return false;
    for (int i = 0; i < aid.length; i++) {
      if (aid[i] != ndefAid[i]) return false;
    }
    return true;
  }
}

File: nfc_proof_of_concept/packages/nfc_host_card_emulation/android/src/main/kotlin/io/flutter/plugins/nfc_host_card_emulation/app_layer/state_machine.kt
package io.flutter.plugins.nfc_host_card_emulation.app_layer

import android.util.Log
import androidx.annotation.GuardedBy
import io.flutter.plugins.nfc_host_card_emulation.file_access.fields.ApduParams
import io.flutter.plugins.nfc_host_card_emulation.file_access.fields.ApduStatusWord
import io.flutter.plugins.nfc_host_card_emulation.file_access.serializers.*
import io.flutter.plugins.nfc_host_card_emulation.ndef_format.NdefMessageSerializer
import io.flutter.plugins.nfc_host_card_emulation.ndef_format.NdefRecordData
import java.lang.Integer.min
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

private enum class HceState {
    IDLE,
    NDEF_APP_SELECTED,
    FILE_SELECTED
}

private class NdefFile(
    message: NdefMessageSerializer,
    val maxFileSize: Int,
    val isWritable: Boolean
) {
    var buffer: Bytes
        private set

    init {
        buffer = Bytes(0)
        update(message)
    }

    fun update(message: NdefMessageSerializer) {
        val messageBytes = message.buffer
        val nlen = messageBytes.size
        require(nlen + 2 <= maxFileSize) { "NDEF message size ($nlen bytes) exceeds max file size ($maxFileSize bytes)." }
        val nlenBytes = Bytes(2)
        nlenBytes[0] = (nlen shr 8 and 0xFF).toByte()
        nlenBytes[1] = (nlen and 0xFF).toByte()
        buffer = nlenBytes + messageBytes
    }

    fun write(offset: Int, data: Bytes): Boolean {
        if (!isWritable) return false
        if (offset < 0 || offset + data.size > maxFileSize) return false

        val newSize = maxOf(buffer.size, offset + data.size)
        val newBuffer = buffer.copyOf(newSize)
        data.copyInto(newBuffer, offset)
        buffer = newBuffer
        return true
    }
}

class HceStateMachine(aid: Bytes) {
    private val stateLock = ReentrantLock()
    private val fileSystemLock = ReentrantLock()
    
    val aid: Bytes = aid.clone() // Defensive copy
    
    @GuardedBy("stateLock")
    private var currentState: HceState = HceState.IDLE
    
    @GuardedBy("stateLock")
    private var selectedFileId: Int? = null

    @GuardedBy("fileSystemLock")
    private var capabilityContainer: CapabilityContainer
    
    @GuardedBy("fileSystemLock")
    private val ndefFiles = mutableMapOf<Int, NdefFile>()
    
    @GuardedBy("fileSystemLock")
    private val fileDescriptors = mutableListOf<FileControlTlv>()

    companion object {
        const val CC_FILE_ID = 0xE103
        const val NDEF_FILE_ID = 0xE104
        
        private val RESERVED_FILE_IDS = setOf(
            0x3F00, 0x3FFF,  // ISO/IEC 7816-4 reserved
            0xE101, 0xE102,  // Reserved by Type 4 Tag spec
            CC_FILE_ID       // CC file
        )
    }

    init {
        capabilityContainer = CapabilityContainer(fileDescriptors = emptyList())
        rebuildCapabilityContainer()
    }

    fun addOrUpdateNdefFile(
        fileId: Int,
        records: List<NdefRecordData>,
        maxFileSize: Int,
        isWritable: Boolean
    ) {
        ValidationUtils.validateFileId(fileId)
        ValidationUtils.validateNdefMessageSize(maxFileSize)

        if (fileId in RESERVED_FILE_IDS) {
            throw InvalidFileIdError("File ID 0x${fileId.toString(16)} is reserved")
        }

        fileSystemLock.withLock {
            try {
                val message = NdefMessageSerializer.fromRecords(records)
                val file = NdefFile(
                    message = message,
                    maxFileSize = maxFileSize,
                    isWritable = isWritable
                )
                ndefFiles[fileId] = file

                val tlv = if (fileId == NDEF_FILE_ID) {
                    FileControlTlv.ndef(
                        maxNdefFileSize = maxFileSize,
                        isNdefWritable = isWritable
                    )
                } else {
                    FileControlTlv.proprietary(
                        proprietaryFileId = fileId,
                        maxProprietaryFileSize = maxFileSize,
                        isProprietaryWritable = isWritable
                    )
                }

                fileDescriptors.removeAll {
                    val id = ((it.fileId.buffer[0].toInt() and 0xFF) shl 8) or 
                            (it.fileId.buffer[1].toInt() and 0xFF)
                    id == fileId
                }
                fileDescriptors.add(tlv)
                rebuildCapabilityContainer()
            } catch (e: Exception) {
                throw InvalidNdefFormatError("Failed to create NDEF file: ${e.message}")
            }
        }
    }

    fun deleteNdefFile(fileId: Int) {
        ValidationUtils.validateFileId(fileId)

        if (fileId in RESERVED_FILE_IDS) {
            throw InvalidFileIdError("Cannot delete reserved file 0x${fileId.toString(16)}")
        }

        fileSystemLock.withLock {
            if (!ndefFiles.containsKey(fileId)) {
                throw FileNotFoundError("File 0x${fileId.toString(16)} does not exist")
            }

            ndefFiles.remove(fileId)
            fileDescriptors.removeAll {
                val id = ((it.fileId.buffer[0].toInt() and 0xFF) shl 8) or 
                        (it.fileId.buffer[1].toInt() and 0xFF)
                id == fileId
            }
            rebuildCapabilityContainer()
        }
    }

    fun clearAllFiles() {
        fileSystemLock.withLock {
            ndefFiles.clear()
            fileDescriptors.clear()
            rebuildCapabilityContainer()
        }
    }

    fun hasFile(fileId: Int): Boolean {
        ValidationUtils.validateFileId(fileId)
        fileSystemLock.withLock {
            return ndefFiles.containsKey(fileId)
        }
    }

    fun onDeactivated() {
        currentState = HceState.IDLE
        selectedFileId = null
    }

    fun processCommand(rawCommand: Bytes): Bytes {
        return try {
            val command = ApduCommand.fromBytes(rawCommand)
            if (command.cla != ApduClass.standard) {
                return ApduResponse.error(ApduStatusWord.claNotSupported).buffer
            }

            stateLock.withLock {
                handleCommand(command).buffer
            }
        } catch (e: HceError) {
            Log.e("HceStateMachine", "HCE Error: ${e.message}")
            when (e) {
                is InvalidStateError -> ApduResponse.error(ApduStatusWord.conditionsNotSatisfied)
                is FileNotFoundError -> ApduResponse.error(ApduStatusWord.fileNotFound)
                is BufferOverflowError -> ApduResponse.error(ApduStatusWord.wrongLength)
                else -> ApduResponse.error(ApduStatusWord.conditionsNotSatisfied)
            }.buffer
        } catch (e: Exception) {
            Log.e("HceStateMachine", "Unexpected error: ${e.message}")
            ApduResponse.error(ApduStatusWord.conditionsNotSatisfied).buffer
        }
    }

    private fun rebuildCapabilityContainer() {
        fileSystemLock.withLock {
            capabilityContainer = CapabilityContainer(fileDescriptors = fileDescriptors.toList())
        }
    }

    private fun handleCommand(command: ApduCommand): ApduResponse {
        // Called with stateLock held
        return when (currentState) {
            HceState.IDLE -> handleIdleState(command)
            HceState.NDEF_APP_SELECTED -> handleNdefAppSelectedState(command)
            HceState.FILE_SELECTED -> handleFileSelectedState(command)
        }
    }

    private fun handleIdleState(command: ApduCommand): ApduResponse {
        if (command is SelectCommand && command.data.buffer.contentEquals(this.aid)) {
            currentState = HceState.NDEF_APP_SELECTED
            selectedFileId = null
            return ApduResponse.success()
        }
        return ApduResponse.error(ApduStatusWord.conditionsNotSatisfied)
    }

    private fun handleNdefAppSelectedState(command: ApduCommand): ApduResponse {
        if (command is SelectCommand) {
            val fileId = ((command.data.buffer[0].toInt() and 0xFF) shl 8) or (command.data.buffer[1].toInt() and 0xFF)
            if (fileId == CC_FILE_ID) {
                currentState = HceState.FILE_SELECTED
                selectedFileId = fileId
                return ApduResponse.success()
            }
            if (ndefFiles.containsKey(fileId)) {
                currentState = HceState.FILE_SELECTED
                selectedFileId = fileId
                return ApduResponse.success()
            }
            return ApduResponse.error(ApduStatusWord.fileNotFound)
        }
        return ApduResponse.error(ApduStatusWord.insNotSupported)
    }

    private fun handleFileSelectedState(command: ApduCommand): ApduResponse {
        if (command is SelectCommand) {
            currentState = HceState.NDEF_APP_SELECTED
            selectedFileId = null
            return handleNdefAppSelectedState(command)
        }

        val currentFileId = selectedFileId ?: return ApduResponse.error(ApduStatusWord.conditionsNotSatisfied)

        return when (command) {
            is ReadBinaryCommand -> processRead(command, currentFileId)
            is UpdateBinaryCommand -> processUpdate(command, currentFileId)
            else -> ApduResponse.error(ApduStatusWord.insNotSupported)
        }
    }

    private fun getFileBuffer(fileId: Int): Bytes? {
        return when (fileId) {
            CC_FILE_ID -> capabilityContainer.buffer
            else -> ndefFiles[fileId]?.buffer
        }
    }

    private fun processRead(command: ReadBinaryCommand, fileId: Int): ApduResponse {
        val fileBuffer = getFileBuffer(fileId) ?: return ApduResponse.error(ApduStatusWord.fileNotFound)
        val offset = command.offset
        if (offset < 0 || offset > fileBuffer.size) {
            return ApduResponse.error(ApduStatusWord.wrongOffset)
        }
        if (offset == fileBuffer.size) {
             return ApduResponse.success(Bytes(0))
        }

        val lengthToRead = if (command.lengthToRead == 0) 256 else command.lengthToRead
        val bytesRemaining = fileBuffer.size - offset
        val bytesToSend = min(lengthToRead, bytesRemaining)

        val chunk = fileBuffer.sliceArray(offset until offset + bytesToSend)
        return ApduResponse.success(chunk)
    }

    private fun processUpdate(command: UpdateBinaryCommand, fileId: Int): ApduResponse {
        if (fileId == CC_FILE_ID) {
            return ApduResponse.error(ApduStatusWord.conditionsNotSatisfied)
        }

        val ndefFile = ndefFiles[fileId] ?: return ApduResponse.error(ApduStatusWord.fileNotFound)
        if (!ndefFile.isWritable) {
            return ApduResponse.error(ApduStatusWord.conditionsNotSatisfied)
        }

        val offset = command.offset
        val dataToWrite = command.dataToWrite
        if (offset < 0 || offset > ndefFile.maxFileSize) {
            return ApduResponse.error(ApduStatusWord.wrongOffset)
        }
        if (offset + dataToWrite.size > ndefFile.maxFileSize) {
            return ApduResponse.error(ApduStatusWord.wrongLength)
        }

        ndefFile.write(offset, dataToWrite)
        return ApduResponse.success()
    }
}